src/main/java/seedu/address/logic/commands/ExpenseCommand.java

```diff
@@ -0,0 +1,152 @@
+    public static final String MESSAGE_MISSING_PERSON_DESCRIPTION =
+            "At least one person is not found in the activity\nNo expense was added";
+
+    private final List<String> persons;
+    private final List<Amount> amounts;
+    private final String description;
+
+    public ExpenseCommand(List<String> persons, List<Amount> amounts, String description) {
+        requireAllNonNull(persons, amounts, description);
+        this.persons = persons;
+        this.amounts = amounts;
+        this.description = description;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
+        if (persons.size() != amounts.size()) {
+            throw new CommandException(MESSAGE_INCONSISTENT_EXPENSES);
+        }
+
+        List<Person> searchScope;
+        Activity activity;
+
+        // Contextual behaviour
+        if (model.getContext().getType() != Context.Type.VIEW_ACTIVITY) {
+            if (!Title.isValidTitle(description)) {
+                throw new CommandException(MESSAGE_MISSING_DESCRIPTION);
+            }
+            activity = new Activity(new Title(description));
+            model.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL_PERSONS);
+        } else {
+            activity = model.getContext().getActivity().get();
+            model.updateFilteredPersonList(x -> activity.getParticipantIds().contains(x.getPrimaryKey()));
+        }
+
+        searchScope = model.getFilteredPersonList();
+
+        // This loop adds expenses one by one using keyword matching.
+        // For each participant argument passed through, the argument is broken up into keywords
+        // which are then used to search through the searchScope (context dependent).
+        // Expenses will only be added if every keyword string has a unique match.
+        StringBuilder successMessage = new StringBuilder();
+        List<String> keywords;
+        List<Person> findResult;
+        Expense[] expenseList = new Expense[persons.size()];
+        for (int i = 0; i < persons.size(); i++) {
+            keywords = Arrays.asList(persons.get(i).split(" "));
+            NameContainsAllKeywordsPredicate predicate = new NameContainsAllKeywordsPredicate(keywords);
+            findResult = searchScope.stream().filter(predicate).collect(Collectors.toList());
+
+            if (findResult.size() != 1) {
+                throw new CommandException(String.format(MESSAGE_NON_UNIQUE_SEARCH_RESULT, persons.get(i)));
+            }
+
+            Person person = findResult.get(0);
+
+            successMessage.append(String.format(MESSAGE_EXPENSE, person.getName(), amounts.get(i), description));
+            expenseList[i] = new Expense(person.getPrimaryKey(), amounts.get(i), description);
+
+            // Contextual behaviour
+            if (model.getContext().getType() != Context.Type.VIEW_ACTIVITY) {
```

Comment:
But the command still needs to dig into the context object to look for the actual object of interest (be it activity or contact or otherwise), unless we redesign the way context / model is represented